web/packages/extension/src/player.ts

```diff
@@ -236,22 +239,120 @@ window.addEventListener("load", () => {
 });
 
 async function loadSwf(swfUrl: string) {
-    try {
-        const pathname = new URL(swfUrl).pathname;
-        document.title = pathname.substring(pathname.lastIndexOf("/") + 1);
-    } catch (_) {
-        // Ignore URL parsing errors.
-    }
+    document.title = swfUrl
+        .split("/")
+        .filter((item) => item !== "")
+        .slice(-1)[0]!;
+
+    const resolveUrlHook = async (swfUrl: string) => {
+        const url = await resolveSwfUrl(swfUrl);
+        if (url !== null) {
+            return url.toString();
+        } else {
+            return swfUrl;
+        }
+    };
 
     const options = await utils.getExplicitOptions();
     localFileName.textContent = document.title;
     localFileInput.value = "";
-    load({
-        ...options,
-        url: swfUrl,
-        base: swfUrl.substring(0, swfUrl.lastIndexOf("/") + 1),
-        ...baseExtensionConfig,
-    });
+    load(
+        {
+            ...options,
+            url: swfUrl,
+            base: swfUrl.substring(0, swfUrl.lastIndexOf("/") + 1),
+            ...baseExtensionConfig,
+        },
+        resolveUrlHook,
+    );
+}
+
+/**
+ * Resolves a given string to a URL if possible.
+ * If the protocol is missing and the string is otherwise a valid web URL, https:// is inserted if the
+ * server supports https, otherwise http.
+ * If the protocol is missing and the string is otherwise a valid file URL, file:/// is inserted.
+ * If the string can't be resolved to a URL, null is returned.
+ * @param enteredUrl The string that should be resolved to a URL.
+ * @return The resolved URL object.
+ */
+async function resolveSwfUrl(enteredUrl: string): Promise<URL | null> {
+    // TODO: Use canParse in the future when it doesn't break browser compatibility
+    // If the URL is (very likely) a file URL with missing file protocol, we return it as file URL
+    // Must be the first test as the URL constructor accepts C:\… as URL with protocol C
+    if (enteredUrl.match(/^[A-Za-z]:\\|^[/~\\]/)) {
+        try {
+            return new URL("file:///" + enteredUrl);
+        } catch (error) {
+            return null;
+        }
+    }
+
+    try {
+        return new URL(enteredUrl);
+    } catch (error) {
+        // The protocol is missing
+
+        // If the URL doesn't contain a dot, it can't be a valid web URL
+        // The URL constructor doesn't check this if a protocol exists
+        if (!enteredUrl.includes(".")) {
+            return null;
+        }
+
+        try {
+            // Only use http if it works but https doesn't
```

Comment:
I am not totally convinced that we shouldn't just default to https:

Reply:
That's because HTTP website cannot fetch HTTPS resources, and vice versa.

Reply:
How is this relevant? The "page" fetching this is our extension.

Reply:
Oh, then that won't be a problem.

Reply:
But what if the website only supports http and not https.

Reply:
I thought about that for a longer time too. But Ruffle is used to play Flash games, mostly legacy content. Old SWF files are especially likely to be hosted on outdated websites which don't support HTTPS yet.
And if the website uses HTTPS, that's just used. HTTP is only used if it's the websites only supports it, but not HTTPS.